reaktive/src/commonMain/kotlin/com/badoo/reaktive/observable/WithLatestFrom.kt

```diff
@@ -0,0 +1,82 @@
+package com.badoo.reaktive.observable
+
+import com.badoo.reaktive.base.ErrorCallback
+import com.badoo.reaktive.completable.CompletableCallbacks
+import com.badoo.reaktive.disposable.CompositeDisposable
+import com.badoo.reaktive.disposable.Disposable
+import com.badoo.reaktive.utils.Uninitialized
+import com.badoo.reaktive.utils.lock.newLock
+import com.badoo.reaktive.utils.lock.synchronized
+import kotlin.jvm.JvmName
+
+fun <T, O, R> Observable<T>.withLatestFrom(
+    other: Observable<O>,
+    combine: (value: T, other: O) -> R
+): Observable<R> =
+    observable { emitter ->
+        val proxy = Proxy(combine, emitter::onNext)
+        val disposables = CompositeDisposable()
+
+        emitter.setDisposable(disposables)
+
+        this@withLatestFrom.subscribeSafe(
+            object : ObservableObserver<T>, CompletableCallbacks by emitter {
+
+                override fun onNext(value: T) = proxy.onNext(value)
+
+                override fun onSubscribe(disposable: Disposable) {
+                    disposables += disposable
+                    other.subscribeSafe(
+                        object : ObservableObserver<O>, ErrorCallback by emitter {
+
+                            override fun onSubscribe(disposable: Disposable) {
+                                disposables += disposable
+                            }
+
+                            override fun onNext(value: O) = proxy.onNext(value)
+
+                            override fun onComplete() {
+                            }
+
+                        }
+                    )
+                }
+            }
+        )
+    }
+
+private class Proxy<T, O, R>(
```

Comment:
Better to name private classes like "WithLatestFromProxy", otherwise you can't have more than one class called `Proxy` in the same package.